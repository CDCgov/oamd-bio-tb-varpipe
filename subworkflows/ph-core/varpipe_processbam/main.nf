include { BWA_INDEX                                                             } from "../../../modules/ph-core/bwa/index/main"
include { VARPIPE_BWAMEM                                                        } from "../../../modules/ph-core/varpipe/bwamem/main"
include { SAMTOOLS_FAIDX                                                        } from "../../../modules/ph-core/samtools/faidx/main"
include { PICARD_SORTSAM                                                        } from "../../../modules/ph-core/picard/sortsam/main"
include { PICARD_MARKDUPLICATES                                                 } from "../../../modules/ph-core/picard/markduplicates/main"
include { PICARD_CREATESEQUENCEDICTIONARY                                       } from "../../../modules/ph-core/picard/createsequencedictionary/main"
include { SAMTOOLS_VIEW as SAMTOOLS_VIEW_UNMAPPED                               } from "../../../modules/ph-core/samtools/view/main"
include { SAMTOOLS_VIEW                                                         } from "../../../modules/ph-core/samtools/view/main"
include { SAMTOOLS_VIEW as SAMTOOLS_VIEW_MAPPED                                 } from "../../../modules/ph-core/samtools/view/main"
include { PICARD_SAMFORMATCONVERTER                                             } from "../../../modules/ph-core/picard/samformatconverter/main"
include { PICARD_BUILDBAMINDEX                                                  } from "../../../modules/ph-core/picard/buildbamindex/main"
include { PICARD_BUILDBAMINDEX as PICARD_BUILDBAMINDEX_2                        } from "../../../modules/ph-core/picard/buildbamindex/main"
include { SAMTOOLS_DEPTH                                                        } from "../../../modules/ph-core/samtools/depth/main"

workflow VARPIPE_PROCESSBAM {
    take:
    ch_reads // channel (mandatory): [ val(meta), path(reads) ]
    ch_ref_fasta // channel (mandatory): [ val(meta), path(ref_fasta) ]

    main:
    ch_versions = Channel.empty()

   // ALIGNERS
    BWA_INDEX (
        ch_ref_fasta
    )
    ch_versions = ch_versions.mix(BWA_INDEX.out.versions)

    PICARD_CREATESEQUENCEDICTIONARY (
        ch_ref_fasta
    )
    ch_versions = ch_versions.mix(PICARD_CREATESEQUENCEDICTIONARY.out.versions)

    SAMTOOLS_FAIDX (
        ch_ref_fasta,
        [[],[]],
        []
    )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX.out.versions)

    ch_index = BWA_INDEX.out.index
        .join(PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict)
        .join(SAMTOOLS_FAIDX.out.fai)
        .join(ch_ref_fasta)
        .collect()

    VARPIPE_BWAMEM (
        ch_reads,
        ch_index
    )
    ch_versions = ch_versions.mix(VARPIPE_BWAMEM.out.versions)

    // CONVERT SAM TO BAM
    PICARD_SAMFORMATCONVERTER (
        VARPIPE_BWAMEM.out.sam
    )
    ch_versions = ch_versions.mix(PICARD_SAMFORMATCONVERTER.out.versions)

    // RUN MAPPING REPORT AND MARK DUPLICATES
    PICARD_SORTSAM (
        PICARD_SAMFORMATCONVERTER.out.bam,
        'coordinate'
    )
    ch_versions = ch_versions.mix(PICARD_SORTSAM.out.versions)

    PICARD_MARKDUPLICATES (
        PICARD_SORTSAM.out.bam,
        ch_ref_fasta,
        SAMTOOLS_FAIDX.out.fai.collect()
    )
    ch_versions = ch_versions.mix(PICARD_MARKDUPLICATES.out.versions)

    PICARD_BUILDBAMINDEX (
        PICARD_MARKDUPLICATES.out.bam
    )
    ch_versions = ch_versions.mix(PICARD_BUILDBAMINDEX.out.versions)

    SAMTOOLS_VIEW_UNMAPPED (
        PICARD_MARKDUPLICATES.out.bam.join(PICARD_BUILDBAMINDEX.out.bai),
        [[],[]],
        [],
        []
    )
    ch_versions = ch_versions.mix(SAMTOOLS_VIEW_UNMAPPED.out.versions)


    SAMTOOLS_VIEW (
        PICARD_MARKDUPLICATES.out.bam.join(PICARD_BUILDBAMINDEX.out.bai),
        [[],[]],
        [],
        []
    )
    ch_versions = ch_versions.mix(SAMTOOLS_VIEW.out.versions)

    PICARD_BUILDBAMINDEX_2 (
        SAMTOOLS_VIEW.out.bam
    )
    ch_versions = ch_versions.mix(PICARD_BUILDBAMINDEX_2.out.versions)

    SAMTOOLS_VIEW_MAPPED (
        SAMTOOLS_VIEW.out.bam.join(PICARD_BUILDBAMINDEX_2.out.bai),
        [[],[]],
        [],
        []
    )
    ch_versions = ch_versions.mix(SAMTOOLS_VIEW_MAPPED.out.versions)

    // RUN GENOME COVERAGE STATISTICS
    SAMTOOLS_DEPTH (
        SAMTOOLS_VIEW.out.bam,
        [[],[]]
    )
    ch_versions = ch_versions.mix(SAMTOOLS_DEPTH.out.versions)

    emit:
    bam                   = SAMTOOLS_VIEW.out.bam // channel (mandatory): [ val(meta), path(bam) ]
    bai                   = PICARD_BUILDBAMINDEX_2.out.bai // channel (mandatory): [ val(meta), path(reads) ]
    coverage              = SAMTOOLS_DEPTH.out.tsv // channel (mandatory): [ val(meta), path(tsv) ]
    fai                   = SAMTOOLS_FAIDX.out.fai // channel (mandatory): [ val(meta), path(fai) ]
    stats                 = SAMTOOLS_DEPTH.out.tsv // channel (mandatory): [ val(meta), path(tsv) ]
    unmapped              = SAMTOOLS_VIEW_UNMAPPED.out.bam // channel (mandatory): [ val(meta), path(bam) ]
    mapped                = SAMTOOLS_VIEW_MAPPED.out.bam // channel (mandatory): [ val(meta), path(bam) ]
    reference_dict        = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict // channel (mandatory): [ val(meta), path(ref_dict) ]
    versions              = ch_versions // channel (mandatory): [ versions.yml ]
}
