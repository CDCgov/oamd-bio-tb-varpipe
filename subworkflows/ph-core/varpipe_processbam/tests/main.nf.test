nextflow_workflow {

    name "Test Subworkflow VARPIPE_PROCESSBAM"
    script "../main.nf"
    config "./nextflow.config"
    workflow "VARPIPE_PROCESSBAM"

    tag "subworkflows"
    tag "subworkflows_phcore"
    tag "subworkflows/varpipe_processbam"
    tag "picard/buildbamindex"
    tag "picard/markduplicates"
    tag "picard/samformatconverter"
    tag "picard/sortsam"
    tag "picard/createsequencedictionary"
    tag "samtools/faidx"
    tag "samtools/depth"
    tag "samtools/view"
    tag "bwa/index"
    tag "varpipe/bwamem"

    test("varpipe-processbam subworkflow test [paired-end]") {

        when {
            params {
                outdir = "$outputDir"
            }
            workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + '/genomics/tuberculosis/fastqs/test_cleaned_read_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + '/genomics/tuberculosis/fastqs/test_cleaned_read_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = Channel.of([
                [ id:'test', single_end:false ],
                file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/ref2.fa', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert (workflow.out.reference_dict).size() > 0 },
                { assert snapshot(
                    workflow.out.bam,
                    workflow.out.bai,
                    workflow.out.coverage,
                    workflow.out.fai,
                    workflow.out.stats,
                    workflow.out.unmapped,
                    workflow.out.mapped,
                    workflow.out.versions
                ).match() }
            )
        }
    }
}
