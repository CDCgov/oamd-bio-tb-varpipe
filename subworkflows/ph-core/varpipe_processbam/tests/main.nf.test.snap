{
    "varpipe-processbam subworkflow test [paired-end]": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_sdrcsm.bam:md5,f562cc57026ae08106532f19bcbd33cc"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_sdrcsm.bai:md5,a426aaf0106f997bdf11f58a3dac165e"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.tsv:md5,8d9da07d55e76f071e677e1279a77a5d"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "ref2.fa.fai:md5,582b23a33b52d218af2cc8498b64ff95"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.tsv:md5,8d9da07d55e76f071e677e1279a77a5d"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "unmapped.bam:md5,116760ced138d18f1795b79cc428e07d"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "mapped.bam:md5,f136094ff245b3ca7c05e86497773002"
                ]
            ],
            [
                "versions.yml:md5,10ae21f4f7a737b20b9574e225aeaa20",
                "versions.yml:md5,350639575c0394c8e78968e0c6007009",
                "versions.yml:md5,45bb2d30395ed69f648c69baeb5dcc2e",
                "versions.yml:md5,4db4be4c0b1621826dcea5cca2ea17c8",
                "versions.yml:md5,6ea3cd10017798ead3d683cc4b7fa43a",
                "versions.yml:md5,79c73343fe03beb01b43e37ad1ce47fc",
                "versions.yml:md5,7beb7a6d6bffb9c62c610e52d408d662",
                "versions.yml:md5,7f23ce94093858f2d0e53abefe10a27f",
                "versions.yml:md5,a7e4cd8225b048870876a878cea36355",
                "versions.yml:md5,beb8d93b21029db660cd52f987463cfe",
                "versions.yml:md5,c403ee43d832ca499cd97aa0ad2eb558",
                "versions.yml:md5,cb02ee201290e0f3905178e6028010d1",
                "versions.yml:md5,cb03c40f024e850a2096af3d53c6d054"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-05-19T11:08:36.147239753"
    }
}
